lib/src/main/java/com/squareup/cycler/RecyclerMutations.kt

```diff
@@ -43,19 +51,73 @@ class MutationExtensionSpec<T : Any> : ExtensionSpec<T> {
   }
 
   /**
-   * Register a callback that will be invoked when an item is removed by swiping it
+   * Allows to define a [RowSpec] (usual `create/bind` mechanism) for a view that will be drawn
+   * below a swiped row. It's created when the swipe is initiated and bind is called with every
+   * movement receiving [SwipeBindData] so:
+   * - the original swiped view can be retouched accorindg to the percentage (alpha for instance)
+   * - the original dataItem, in case its data is needed for the underview, and the underview itself.
+   * WARNING: this API might evolve as currently:
+   * - under view is only defined at top level, and cannot be customized for each row-type,
+   *   which should be the case using Cycler.
+   * - probably there should be extra callbacks or customizations available for swiping. For
+   *   instance: to restore the original alpha of the swiped view.
    */
-  fun onSwipeToRemove(block: (T) -> Unit) {
-    onSwipeToRemove = block
+  inline fun <reified S : T, V : View> swipeUnderViewSpec(
+    block: StandardRowSpec<Any, SwipeBindData<S>, V>.() -> Unit
+  ) {
+    // RowSpec's super type is Any. We cannot use the proper T
+    // because we are not passing the proper S, but SwipeBindData<...>.
+    swipeUnderViewSpecs += StandardRowSpec<Any, SwipeBindData<S>, V> { it is S }.apply(block)
+  }
+
+  @Deprecated("Use onSwiped", ReplaceWith("onSwiped(block)"))
+  inline fun onSwipeToRemove(crossinline block: (T) -> Unit) {
+    onSwiped { item: T, _: Int -> block(item) }
   }
 
   /**
-   * Register a check that will be invoked for each item to decide whether that item can be swiped to
-   * remove it or not. This allows a hook for recyclers to specify which items do and do not support
-   * swipe to remove.
+   * Register a callback that will be invoked when an item is removed by swiping it.
    */
-  fun canSwipeToRemoveItem(block: (T) -> Boolean) {
-    canSwipeToRemoveItem = block
+  inline fun onSwiped(crossinline block: (T) -> Unit) {
+    onSwiped { item: T, _: Int -> block(item) }
+  }
+
+  @Deprecated("Use onSwiped", ReplaceWith("onSwiped(block)"))
+  fun onSwipeToRemove(block: (T, direction: Int) -> Unit) {
+    onSwiped(block)
+  }
+
+  /**
+   * Register a callback that will be invoked when an item is removed by swiping it.
+   * @param block receives the swiped data item and the direction. Direction is one of
+   * [ItemTouchHelper.START] or [ItemTouchHelper.END].
+   */
+  fun onSwiped(block: (T, direction: Int) -> Unit) {
+    onSwiped = block
+  }
+
+  @Deprecated(
+    message = "Use the version that specifies which side is allowed.",
+    replaceWith = ReplaceWith(
+      expression = "canSwipe { if(block(it)) ItemTouchHelper.START or ItemTouchHelper.END else 0 }"
```

Comment:
I didn't add the `+` because in some cases somebody will want to return only one side and they'll have to manually convert to set, or do something special.

I tried to define two versions of the method, one for `block: () -> SD` another for `block: () -> Set<SD>` but the type erasure ends up receiving always `KFunction1`, so I cannot provide that.

If you have an idea for that, I'll be glad to include.

Reply:
Asking people to type `setOf(ItemTouchHelper.START)` doesn't seem that bad.

But you could also just make a class that has `start` and `end` booleans, and either just provide that directly (`if (block(it)) Direction(start = true, end = true) else Direction.None`) or add some sugar like this (probably overkill:

```kotlin
abstract class Flags private constructor(
  val isStart: Boolean = false,
  val isEnd: Boolean = false
) {
  object Start : Flags(isStart = true)
  object End : Flags(isEnd = true)

  operator fun plus(otherFlags: Flags): Flags = object : Flags(
      isStart = isStart || otherFlags.isStart,
      isEnd = isEnd || otherFlags.isEnd
  ) {}
}
```

Reply:
Looks like an EnumSet :)
Maybe I can even define an `inline SwipeDirections(val enumSet: ...)` with secondary constructors that build the right thing.